src/middle/Stmt.ml

```diff
@@ -263,18 +272,83 @@ module Helpers = struct
     let rec go st bodyfn var smeta =
       match st with
       | SizedType.SArray (t, d) ->
-          let bodyfn' var = mk_for_iteratee d bodyfn var smeta in
+          let bodyfn' _ var = mk_for_iteratee d (bodyfn st) var smeta in
           go t bodyfn' var smeta
       | SMatrix (mem_pattern, d1, d2) ->
-          let bodyfn' var = mk_for_iteratee d1 bodyfn var smeta in
+          let bodyfn' _ var = mk_for_iteratee d1 (bodyfn st) var smeta in
           go (SRowVector (mem_pattern, d2)) bodyfn' var smeta
       | SComplexMatrix (d1, d2) ->
-          let bodyfn' var = mk_for_iteratee d1 bodyfn var smeta in
+          let bodyfn' _ var = mk_for_iteratee d1 (bodyfn st) var smeta in
           go (SComplexRowVector d2) bodyfn' var smeta
       | _ -> for_scalar st bodyfn var smeta in
-    go st (Fn.compose bodyfn invert_index_order) var smeta
+    go st (fun st var -> bodyfn st (invert_index_order var)) var smeta
 
-  let assign_indexed decl_type vident meta varfn var =
+  let assign_indexed decl_type lval meta varfn var =
     let indices = Expr.Helpers.collect_indices var in
-    Fixed.{meta; pattern= Assignment ((vident, decl_type, indices), varfn var)}
+    Fixed.
+      { meta
+      ; pattern= Assignment (LIndexed (lval, indices), decl_type, varfn var) }
+
+  let rec get_lhs_name (lval : 'a Fixed.Pattern.lvalue) =
+    match lval with
+    | LVariable name -> name
+    | LIndexed (sub_lval, _) -> get_lhs_name sub_lval
+    | LTupleProjection (sub_lval, num) ->
+        get_lhs_name sub_lval ^ "." ^ string_of_int num
+
+  (* Copied from AST's version in AST.ml *)
+  let rec lvalue_of_expr_opt (expr : 'e Expr.Fixed.t) :
+      'e Expr.Fixed.t Fixed.Pattern.lvalue option =
+    match expr.pattern with
+    | Var s -> Some (LVariable s)
+    | Indexed (l, i) ->
+        Option.bind (lvalue_of_expr_opt l) ~f:(fun lv ->
+            Some (Fixed.Pattern.LIndexed (lv, i)) )
```

Comment:
Yeah, this should be `Option.map` instead of `bind`, so it will look like 
```ocaml
          Option.map (lvalue_of_expr_opt l) ~f:(fun lv ->
              Fixed.Pattern.LTupleProjection (lv, i) )
```

One fewer sets of nesting required same meaning